akka-stream/src/main/scala/akka/stream/impl/Throttle.scala

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import akka.stream.ThrottleMode.{ EnforcingMode, ShapingMode }
+import akka.stream.impl.fusing.GraphStages.SimpleLinearGraphStage
+import akka.stream.stage._
+import akka.stream._
+
+import scala.concurrent.duration.{ FiniteDuration, _ }
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ */
+private[stream] trait ThrottleTimeCalculationHelper {
+  def now(): Long = System.currentTimeMillis()
+
+  def calculateSpeed(counter: Double, time: FiniteDuration): Long =
+    ((counter / time.toMillis) * 1024 * 1024).toLong
+
+  def scale(e: Int): Long = e.toLong << 20
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] abstract class AbstractThrottle[T](cost: Int,
+                                                   per: FiniteDuration,
+                                                   maximumBurst: Int,
+                                                   mode: ThrottleMode)
+  extends SimpleLinearGraphStage[T] with ThrottleTimeCalculationHelper {
+
+  def calculateElementCost(elem: T): Long
+  val timerName: String
+
+  val enforcedModeException = new IllegalStateException("Maximum throttle throughput exceeded")
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
+    var willStop = false
+    var lastTokens: Long = 0
+    var previousTime: Long = 0
+
+    var currentElement: Option[T] = None
+
+    setHandler(in, new InHandler {
+      val scaledMaximumBurst = scale(maximumBurst)
+
+      override def onUpstreamFinish(): Unit =
+        if (isAvailable(out) && isTimerActive(timerName)) willStop = true
+        else completeStage()
+
+      override def onPush(): Unit = {
+        val timeElapsed = now() - previousTime
+        val speed = calculateSpeed(cost, per)
+        val currentTokens = Math.min(timeElapsed * speed + lastTokens, scaledMaximumBurst)
+        val elem = grab(in)
+        try {
+          val elementCost = calculateElementCost(elem)
+          if (currentTokens < elementCost)
+            mode match {
+              case ShapingMode ⇒
+                currentElement = Some(elem)
+                scheduleOnce(timerName, ((elementCost - currentTokens) / speed).millis)
+              case EnforcingMode ⇒ failStage(enforcedModeException)
+            }
+          else {
+            lastTokens = currentTokens - elementCost
+            previousTime = now()
+            push(out, elem)
+          }
+        } catch { case NonFatal(ex) ⇒ failStage(ex) } // for calculatingCost
```

Comment:
no need, the stage environment will automatically fail the stage on escaped exceptions.